File: client/rz_mesh.js
/*
    This file is part of rhizi, a collaborative knowledge graph editor.
    Copyright (C) 2014-2015  Rhizi

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published
    by the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/**
 * Manage backend websocket connection
 */
define('rz_mesh',
       [ 'util', 'model/diff', 'model/util', 'socketio', 'view/activity'],
function( util,   model_diff,   model_util,   io,              activity) {

    "use strict";

    var ws_server_url = document.location.origin + '/graph'; // socketio namespace

    var socket;
    var rz_mesh_graph_ref;

    /**
     * [!] caller is responsible of calling destroy()
     *
     * @param init_spec: expected to contain a graph:graph mapping
     */
    function init(init_spec) {

        socket = io.connect(ws_server_url, {
            'reconnectionDelay': 30000,
        });

        util.assert(undefined !== init_spec.graph, 'unable to init ws connection, graph undefined');
        rz_mesh_graph_ref = init_spec.graph;

        // wire up event handlers
        socket.on('connect', on_connect);
        socket.on('disconnect', on_disconnect);
        socket.on('error', on_error);
        socket.on('diff_commit__topo', ws_diff_merge__topo);
        socket.on('diff_commit__attr', ws_diff_merge__attr);
    }

    function destroy(init_spec) {
        socket.disconnect();
        console.log('ws: connection closed on \'beforeunload\' event'); // no one will ever see this but still
    }

    function on_connect() {
        console.log('ws: connection established: endpoint: ' + ws_server_url);
    }

    function on_disconnect() {
        console.log('ws: connection closed on peer disconnect, endpoint: ' + ws_server_url);
    }

    function on_error(err) {
        console.log('ws: error: ' + err);
    }

    /**
     * Handle websocket incoming Topo_Diffs
     *
     * @param topo_diff_cr: Topo_Diff commit result object
     */
    function ws_diff_merge__topo(topo_diff_spec_raw, topo_diff_cr) {

        // adapt from wire format, no need to do the same for id_sets
        var node_set_add = topo_diff_spec_raw.node_set_add.map(model_util.adapt_format_read_node);
        var link_ptr_set = topo_diff_spec_raw.link_set_add.map(model_util.adapt_format_read_link_ptr);

        var topo_diff_spec = { node_set_add: node_set_add,
                               link_set_add: link_ptr_set,
                               node_id_set_rm: topo_diff_spec_raw.node_id_set_rm,
                               link_id_set_rm: topo_diff_spec_raw.link_id_set_rm,
                               meta: topo_diff_spec_raw.meta };

        // [!] note: this is not a pure Topo_Diff object in the sense it contain a link_ptr_set,
        // not a resolved link object set
        var topo_diff = model_diff.new_topo_diff(topo_diff_spec); // run through validation

        console.log('ws: rx: ws_diff_merge__topo, committing wire-adapted topo_diff:', topo_diff);

        activity.incomingActivityBus.push(topo_diff_spec);
        rz_mesh_graph_ref.commit_diff__topo(topo_diff);
    }

    /**
     * Handle websocket incoming Attr_Diffs
     *
     * @param attr_diff_cr: Attr_Diff commit result object
     */
    function ws_diff_merge__attr(attr_diff_spec, attr_diff_cr) {
        var attr_diff = model_diff.new_attr_diff_from_spec(attr_diff_spec); // run through validation
        console.log('ws: rx: ws_diff_merge__attr, committing wire-adapted topo_diff:', attr_diff);

        activity.incomingActivityBus.push(attr_diff_spec);
        rz_mesh_graph_ref.commit_diff__attr(attr_diff);
    }

    function emit__rzdoc_subscribe(rzdoc_name) {
        var callback = function(data) {
            console.log('ws: rzdoc subscription acked: rzdoc name: \'' + rzdoc_name + '\'');
        };
        socket.emit('rzdoc_subscribe', { rzdoc_name: rzdoc_name}, callback);
    }

    function emit__rzdoc_unsubscribe(rzdoc_name) {
        var callback = function(data) {
            console.log('ws: rzdoc un-subscription acked: rzdoc name: \'' + rzdoc_name + '\'');
        };
        socket.emit('rzdoc_unsubscribe', { rzdoc_name: rzdoc_name}, callback);
    }

    return {
        init : init,
        destroy: destroy,
        emit__rzdoc_subscribe: emit__rzdoc_subscribe,
        emit__rzdoc_unsubscribe: emit__rzdoc_unsubscribe,
    };

});

File: rhizi/rz_mesh.py
#    This file is part of rhizi, a collaborative knowledge graph editor.
#    Copyright (C) 2014-2015  Rhizi
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU Affero General Public License as published
#    by the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU Affero General Public License for more details.
#
#    You should have received a copy of the GNU Affero General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.


from collections import namedtuple
from functools import wraps
import inspect
import logging

from flask import request
import socketio

from .model.graph import Topo_Diff
from .rz_api_rest import Req_Context
from .rz_api_websocket import WebSocket_Graph_NS
from .rz_req_handling import (make_response__http__empty,
    HTTP_STATUS__500_INTERNAL_SERVER_ERROR, HTTP_STATUS__200_OK,
    make_response__json, sock_addr_from_env_HTTP_headers,
    sock_addr_from_REMOTE_X_keys)


log = logging.getLogger('rhizi')


def init_ws_interface(cfg, kernel, flask_webapp):
    """
    Initialize websocket interface:
       - apply websocket route handlers

    @return: an initialized RZ_WebSocket_Server object
    """

    def decorator__ws_multicast(sio, f):
        """
        Emit multicast on topo_diff, attr_diff

        @param f: [!] wrapped function, name used to derive socket message name
        """

        def _prep_for_serialization(obj):
            # handle special serialization cases
            if isinstance(obj, Topo_Diff):
                return obj.to_json_dict()
            return obj

        @wraps(f)
        def wrapped_function(*args, **kw):

            f_ret = f(*args, **kw)
            assert type(f_ret) in [list, tuple]  # expect (X_Diff, X_Diff.Commit_Result_Type)

            diff = f_ret[1]

            data = _prep_for_serialization(diff)

            is_topo_diff = isinstance(f_ret[0], Topo_Diff)

            if is_topo_diff:
                rzdocs = [rzdoc_from_f_args_extractor(args)]
            else:
                node_ids = diff.node_ids()
                link_ids = diff.link_ids()
                rzdocs = kernel.rzdoc__rzdocs_from_ids(node_ids=node_ids, link_ids=link_ids)

            event = f.__name__
            websocket_graph_ns.broadcast_to_rzdocs_readers(event=event, data=data,
                                                           rzdocs=rzdocs)

            return f_ret

        return wrapped_function

    def rzdoc_from_f_args_extractor(f_args):  # extract rzdoc from req ctx
        if len(f_args) < 2 or not isinstance(f_args[1], Req_Context):
            return
        req_ctx = f_args[1]
        return req_ctx.rzdoc

    # connect socketio route
    sio = socketio.Server(namespace='/graph', async_mode='gevent')
    ret_flask_webapp = socketio.Middleware(sio, flask_webapp)

    class Unwrapped(object):
        diff_commit__topo = kernel.diff_commit__topo
        diff_commit__attr = kernel.diff_commit__attr
    kernel.unwrapped = Unwrapped

    kernel.diff_commit__topo = decorator__ws_multicast(sio,
                                                       kernel.diff_commit__topo)
    kernel.diff_commit__attr = decorator__ws_multicast(sio,
                                                       kernel.diff_commit__attr)

    ret_flask_webapp.kernel = kernel
    ret_flask_webapp.rz_websocket = websocket_graph_ns = WebSocket_Graph_NS(sio=sio, kernel=kernel)
    return ret_flask_webapp
